# Segmenting and Clustering Neighborhoods in Toronto

# 1st PART:-  Scraping Data from Wikipedia and Converting it into a Pandas Dataframe

In [1]:
from bs4 import BeautifulSoup as bsoup
from urllib.request import urlopen as uReq
import requests
import lxml
import pandas as pd
from pandas import DataFrame
import numpy as np

In [2]:
wiki_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

In [3]:
s=requests.get(wiki_url)

# By Using the BeautifulSoup Package, We will parse the html File 

In [4]:
wiki_page=bsoup(s.text,"html.parser")
wiki_page

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgMonthNamesShort":["","Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],"wgRequestId":"XjtoiQpAAEEAAD59DRQAAAAK","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":935851093,"wgRevisionId":935851093,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontari

# As we can see from the html file that there is a lot of information related to the neighbourhoods of toronto. Also, we can observe that there is a table class associated with the file, that is, "wikitable sortable" 

In [5]:
wiki_table=wiki_page.table
wiki_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Harbourfront</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Lawrence_Heights" title="Lawrence Heights">Lawrence Heights</a>
</td></tr>
<tr>
<td>M6A</td>
<td><a href="/wiki/North

# Since there is a lot of data present in the file, Let's find the actual length of the data in the form of total rows 

In [6]:
results=wiki_table.find_all('tr')
wiki_rows=len(results)
wiki_rows

288

# So, We have a 'Results' Table which contains 288 rows in which the first row is the table header, so we have 287 rows which contains the data.

In [7]:
results[0:5]

[<tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighbourhood
 </th></tr>, <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>, <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>, <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>]

# Let's Define the Header of the Dataframe,which is the first row of the table.

In [8]:
header=results[0].text.split()
header

['Postcode', 'Borough', 'Neighbourhood']

# Therefore, The Header consists of the following variables:-  
# 1) Postcode
# 2) Borough 
# 3) Neighbourhood


# Before We Create the Dataframe, We should Check the Rows of the Table. For Instance, Let's Check Row 78.  

# Row Number 78

In [9]:
results[78].text

'\nM2J\nNorth York\nHenry Farm\n'

In [10]:
results[78].text.split('\n')

['', 'M2J', 'North York', 'Henry Farm', '']

In [11]:
Postcode=results[78].text.split('\n')[1] # Splitting the text on Index Number 1, that is, M5J  
Postcode

'M2J'

In [12]:
Borough=results[78].text.split('\n')[2] # Splitting the text on Index Number 2, that is,North York 
Borough

'North York'

In [13]:
Neighbourhood=results[78].text.split('\n')[3] # Splitting the text on Index Number 3, that is,Henry Farm 
Neighbourhood

'Henry Farm'

# We Create a Loop which will extract the Data into a Pandas Dataframe.

In [14]:
records =[]
n=1
while n < wiki_rows :
    Postcode=results[n].text.split('\n')[1]
    Borough=results[n].text.split('\n')[2]
    Neighborhood=results[n].text.split('\n')[3]
    records.append((Postcode, Borough,Neighborhood))
    n=n+1

df=pd.DataFrame(records, columns=['PostalCode', 'Borough', 'Neighbourhood'])
df.head(5)

PostalCode           Borough     Neighbourhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [15]:
df.shape

(287, 3)

# Let's Examine the Last 5 Rows of Data to Ensure that we Have All the Data. 

In [16]:
df.tail()

PostalCode       Borough          Neighbourhood
282        M8Z     Etobicoke              Mimico NW
283        M8Z     Etobicoke     The Queensway West
284        M8Z     Etobicoke  Royal York South West
285        M8Z     Etobicoke         South of Bloor
286        M9Z  Not assigned           Not assigned

# Let's Remove the Boroughs which are 'Not Assigned'

In [17]:
df[df['Borough']=='Not assigned'].count()

PostalCode       77
Borough          77
Neighbourhood    77
dtype: int64

In [18]:
df1=df[~df.Borough.str.contains("Not assigned")]
df1=df1.reset_index(drop=True)
df1.head()

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

In [19]:
df1.shape

(210, 3)

In [20]:
df1.loc[df1['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df1['Borough']
df1.head()

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

In [21]:
df1.shape

(210, 3)

# We Assign Unique PostalCodes To the DataFrame and Evaluate the Number of Unique PostalCodes, Boroughs and Neigbourhoods.

In [22]:
postalcodes = df1['PostalCode'].nunique()
boroughs = df1['Borough'].nunique()
neighbourhoods= df1['Neighbourhood'].nunique()
print('Unique Postalcodes : ' + str(postalcodes))
print('Unique Boroughs  : '+ str(boroughs))
print('Unique Neighbourhoods  :' + str(neighbourhoods))

Unique Postalcodes : 103
Unique Boroughs  : 11
Unique Neighbourhoods  :207


# Also, We Analyze the Neighbourhoods having Similar Postcodes and Merge them.

In [23]:
df1.head()

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

In [24]:
wiki_rows1=len(df1)
wiki_rows1

210

In [25]:
df2=df1 # I Created df2 because while using df1 some Borough Values returned "NaN" values which made the DataFrame Inconsistent. 
df2.head()

PostalCode           Borough     Neighbourhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

# By using two Pointers, one for df2 and another for df1, namely m and n, It will make the process of assigning unique postalcodes Easier and Optimal.
# m will vary from 1 to 210 and n will vary from 0 to 209.

In [26]:
wiki_rows2=len(df2)-1
wiki_rows2

209

In [27]:
n=0

while n < wiki_rows2 :
    post1=df2.iloc[n,0]
    #post1
    m=n+1
    post2=df2.iloc[m,0]
    #post2
    neigh1=df2.iloc[n,2]
    neigh2=df2.iloc[m,2]
    if post1==post2:
        df2.Neighbourhood[n,2] = neigh1=neigh1+','+neigh2
        #df2 = df2[df2.Neighbourhood != 'neigh2']
        df2=df2.drop(df2.index[m])
        wiki_rows2=wiki_rows2-1
        df2 = df2.reset_index(drop=True)
    else:
        n=n+1
df2.index

RangeIndex(start=0, stop=103, step=1)

In [28]:
df2.head()

PostalCode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A  Downtown Toronto   

                                       Neighbourhood  
0                                          Parkwoods  
1                                   Victoria Village  
2  Kingsway Park South West,Mimico NW,The Queensw...  
3                    Lawrence Heights,Lawrence Manor  
4                                       Queen's Park

# As we can see, The Neighbourhoods are merged having the same Postcode.

In [29]:
df2.head(15)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A      Queen's Park   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   
12        M1C       Scarborough   
13        M3C        North York   
14        M4C         East York   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2   Kingsway Park South West,Mimico NW,The Queensw...  
3                     Lawrence Heights,Lawrence Manor  
4                                        Queen's Park  
5                                        Queen's Park  
6                                       Rouge,Malvern  
7                                     Don Mills North  
8                      Woodbine Gardens,Parkview Hill  
9                             Ryerson,Garden District  
10                                          Glencairn  
11  Cloverdale,Islington,Martin Grove,Princess Gar...  
12               Highland Creek,Rouge Hill,Port Union  
13                    Flemingdon Park,Don Mills South  
14                                   Woodbine Heights

In [30]:
df2.tail(15)

PostalCode           Borough  \
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  
88          Humber Bay Shores,Mimico South,New Toronto  
89   Albion Gardens,Beaumond Heights,Humbergate,Jam...  
90                                     L'Amoreaux West  
91                                            Rosedale  
92                     Stn A PO Boxes 25 The Esplanade  
93                               Alderwood,Long Branch  
94                                           Northwest  
95                                         Upper Rouge  
96                          Cabbagetown,St. James Town  
97               First Canadian Place,Underground city  
98         The Kingsway,Montgomery Road,Old Mill North  
99                                Church and Wellesley  
100  Business Reply Mail Processing Centre 969 Eastern  
101  Humber Bay,King's Mill Park,Kingsway Park Sout...  
102  Kingsway Park South West,Mimico NW,The Queensw...

In [31]:
df2.shape

(103, 3)

# Since The Number of Unique PostalCodes are 103, The DataFrame is Accurate.    

# PART 1 ENDS HERE

# PART 2:- Obtaining the Latitude and Longtitude Coordinates of each neighbourhood

# As per the Instruction, I am using the CSV File that has the Geographical Coordinates of each Neighbourhood since the Geocoder Package was inconsistent. 

In [32]:
url='http://cocl.us/Geospatial_data'
df_postalcodes=pd.read_csv(url)
df_postalcodes.head()
df_postalcodes.columns = ['PostalCode', 'Latitude', 'Longitude']# Renaming the 'Postal Code' Column to PostalCode
df_postalcodes.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

# By Sorting the dataframe, that is, df2, on the basis of postalcodes, we can combine it with df_postalcodes with the PostalCodes column being common.

In [33]:
df2_sorted=df2.sort_values('PostalCode')

In [34]:
neighborhoods=pd.merge(df2_sorted,df_postalcodes, how='right', on = 'PostalCode')
neighborhoods.head(20)

PostalCode      Borough                                      Neighbourhood  \
0         M1B  Scarborough                                      Rouge,Malvern   
1         M1C  Scarborough               Highland Creek,Rouge Hill,Port Union   
2         M1E  Scarborough                    Guildwood,Morningside,West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park   
7         M1L  Scarborough                      Clairlea,Golden Mile,Oakridge   
8         M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West   
9         M1N  Scarborough                         Birch Cliff,Cliffside West   
10        M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...   
11        M1R  Scarborough                                   Maryvale,Wexford   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter   
14        M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...   
15        M1W  Scarborough                                    L'Amoreaux West   
16        M1X  Scarborough                                        Upper Rouge   
17        M2H   North York                                  Hillcrest Village   
18        M2J   North York                         Fairview,Henry Farm,Oriole   
19        M2K   North York                                    Bayview Village   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577  
15  43.799525 -79.318389  
16  43.836125 -79.205636  
17  43.803762 -79.363452  
18  43.778517 -79.346556  
19  43.786947 -79.385975

In [35]:
neighborhoods.shape

(103, 5)

# PART 2 Ends Here

# PART 3:- Exploring and Clustering the Neighborhoods in Toronto

In [36]:
conda install -c anaconda pandas-datareader

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\ProgramData\Anaconda3

  added / updated specs:
    - pandas-datareader


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    conda-4.8.2                |           py37_0         3.1 MB  anaconda
    ------------------------------------------------------------
                                           Total:         3.1 MB

The following packages will be SUPERSEDED by a higher-priority channel:

  conda                                         conda-forge --> anaconda




conda-4.8.2          | 3.1 MB    |            |   0% 
conda-4.8.2          | 3.1 MB    |            |   1% 
conda-4.8.2          | 3.1 MB    | 8          |   8% 
conda-4.8.2          | 3.1 MB    | #6         |  17% 
conda-4.8.2          | 3.1 MB    | ##3        |  23% 
conda-4.8.2          | 3.1 MB    | ##7        |  28% 
conda-4.8.2   


EnvironmentNotWritableError: The current user does not have write permissions to the target environment.
  environment location: C:\ProgramData\Anaconda3




In [37]:
import numpy as np 
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json 
!conda install -c conda-forge geopy --yes 
import requests 
from pandas.io.json import json_normalize 
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
print('Libraries imported.')

Solving environment: ...working... 
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::openssl-1.1.1d-he774522_2, defaults/win-64::certifi-2019.9.11-py37_0
  - anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::openssl-1.1.1d-he774522_2, defaults/win-64::ca-certificates-2019.8.28-0
  - anaconda/win-64::openssl-1.1.1d-he774522_2, defaults/win-64::ca-certificates-2019.8.28-0, defaults/win-64::certifi-2019.9.11-py37_0
  - anaconda/win-64::certifi-2019.9.11-py37_0, defaults/win-64::ca-certificates-2019.8.28-0, defaults/win-64::openssl-1.1.1d-he774522_2
  - defaults/win-64::ca-certificates-2019.8.28-0, defaults/win-64::certifi-2019.9.11-py37_0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda

In [38]:
conda install -c conda-forge/label/gcc7 folium

Solving environment: ...working... 
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::certifi-2019.9.11-py37_0, anaconda/win-64::openssl-1.1.1d-he774522_2, defaults/win-64::ca-certificates-2019.8.28-0
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::certifi-2019.9.11-py37_0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::certifi-2019.9.11-py37_0, defaults/win-64::ca-certificates-2019.8.28-0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::ca-certificates-2019.8.28-0, defaults/win-64::certifi-2019.9.11-py37_0, defaults/win-64::openssl-1.1.1d-he774522_2
  - defaults/win-64::ca-certificates-2019.8.28-0, defaults/win-64::certifi-2019.9.11-py37_0, defaults/win-64::openssl-1.1.1d-he774522_2
  - anaconda/win-64::ca-certificates-2019.8.28-0, anaconda/win-64::openssl-1.1.1d-he774522_2, defaults/win-64::certifi-2019.9.11-py37_0
  - anaconda

In [39]:
import folium

# Let's Focus on the Boroughs which are present in Toronto. 

In [40]:
toronto_can= neighborhoods[neighborhoods['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
toronto_can.head()

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [41]:
latitude = 43.6532
longitude= -79.3832

In [42]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_can['Latitude'], toronto_can['Longitude'], toronto_can['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
    ).add_to(map_toronto)  
    
map_toronto

# Let's start using the Foursquare API by defining the Client ID, Client Secret, Version and other parameters.  

In [43]:
CLIENT_ID = 'BV4EOHUQT3AJKYAN2PVQCZJSOKYIH23FZLF3U1F0IVA3STUA' # Foursquare ID
CLIENT_SECRET = 'WK5FS4GDRT5S10PWSKNGINNLO4VE3K4HNCLKQGV0QYVMSRIL' #Foursquare Secret
VERSION = '20200125'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: BV4EOHUQT3AJKYAN2PVQCZJSOKYIH23FZLF3U1F0IVA3STUA
CLIENT_SECRET:WK5FS4GDRT5S10PWSKNGINNLO4VE3K4HNCLKQGV0QYVMSRIL


# I will be using the same approach which was present in the Clustering NewYork Lab Example.

In [44]:
LIMIT = 100 # It will give 100 results which will be returned by the Foursquare API 
radius = 500

In [45]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

# Now By Using this function, I will create a dataframe which will give all the venues present in Toronto.

In [47]:
torontocan_venues = getNearbyVenues(names=toronto_can['Neighbourhood'],
                                   latitudes=toronto_can['Latitude'],
                                   longitudes=toronto_can['Longitude']
                                  )

The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Kingsway Park South West,Mimico NW,The Queensway West,Royal York South West,South of Bloor
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place

In [48]:
torontocan_venues.shape

(1706, 7)

In [49]:
torontocan_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                      Domino's Pizza       43.679058       -79.297382   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3        Pizza Place  
4       Neighborhood

# Let's Check the venues on the basis of their Neighbourhoods.

In [50]:
torontocan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            54   
Brockton,Exhibition Place,Parkdale Village                             23   
Business Reply Mail Processing Centre 969 Eastern                      18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     16   
Cabbagetown,St. James Town                                             41   
Central Bay Street                                                     85   
Chinatown,Grange Park,Kensington Market                                84   
Christie                                                               18   
Church and Wellesley                                                   82   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             37   
Davisville North                                                        6   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     14   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            20   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          36   
Harbourfront East,Toronto Islands,Union Station                       100   
High Park,The Junction South                                           24   
Kingsway Park South West,Mimico NW,The Queenswa...                     45   
Lawrence Park                                                           4   
Little Portugal,Trinity                                                51   
Moore Park,Summerhill East                                              2   
North Toronto West                                                     23   
Parkdale,Roncesvalles                                                  14   
Queen's Park                                                           39   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede,Swansea                                                      35   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        95   
Studio District                                                        41   
The Annex,North Midtown,Yorkville                                      22   
The Beaches                                                             5   
The Beaches West,India Bazaar                                          19   
The Danforth West,Riverdale                                            42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             54   
Brockton,Exhibition Place,Parkdale Village                              23   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      16   
Cabbagetown,St. James Town                                              41   
Central Bay Street                                                      85   
Chinatown,Grange Park,Kensington Market                                 84   
Christie                                                                18   
Church and Wellesley                

# Now I will Analyze each Neighbourhood and also use one-hot encoding.

In [54]:
toronto_onehotencoding = pd.get_dummies(torontocan_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehotencoding['Neighborhood'] = torontocan_venues['Neighborhood'] 
cols=list(toronto_onehotencoding.columns.values)
cols.pop(cols.index('Neighborhood'))
toronto_onehotencoding=toronto_onehotencoding[['Neighborhood']+cols]
toronto_onehotencoding.rename(columns = {'Neighborhood': 'Neighbourhood'}, inplace = True) # for consistent output 
toronto_onehotencoding.head()

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0   The Beaches                  0        0                   0             0   
1   The Beaches                  0        0                   0             0   
2   The Beaches                  0        0                   0             0   
3   The Beaches                  0        0                   0             0   
4   The Beaches                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0        0     0                          0             0            0   
1        0     0                          0             0            0   
2        0     0                          0             0            0   
3        0    

In [55]:
toronto_onehotencoding.shape

(1706, 229)

# Grouping the Neighbourhoods and taking their respective mean.

In [58]:
torontocan_grouping = toronto_onehotencoding.groupby('Neighbourhood').mean().reset_index()
torontocan_grouping


Neighbourhood  Afghan Restaurant  \
0                              Adelaide,King,Richmond           0.000000   
1                                         Berczy Park           0.000000   
2          Brockton,Exhibition Place,Parkdale Village           0.000000   
3   Business Reply Mail Processing Centre 969 Eastern           0.000000   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...           0.000000   
5                          Cabbagetown,St. James Town           0.000000   
6                                  Central Bay Street           0.000000   
7             Chinatown,Grange Park,Kensington Market           0.000000   
8                                            Christie           0.000000   
9                                Church and Wellesley           0.012195   
10                      Commerce Court,Victoria Hotel           0.000000   
11                                         Davisville           0.000000   
12                                   Davisville North           0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...           0.000000   
14            Design Exchange,Toronto Dominion Centre           0.000000   
15                        Dovercourt Village,Dufferin           0.000000   
16              First Canadian Place,Underground city           0.000000   
17                 Forest Hill North,Forest Hill West           0.000000   
18                      Harbord,University of Toronto           0.000000   
19    Harbourfront East,Toronto Islands,Union Station           0.000000   
20                       High Park,The Junction South           0.000000   
21  Kingsway Park South West,Mimico NW,The Queensw...           0.000000   
22                                      Lawrence Park           0.000000   
23                            Little Portugal,Trinity           0.000000   
24                         Moore Park,Summerhill East           0.000000   
25                                 North Toronto West           0.000000   
26                              Parkdale,Roncesvalles           0.000000   
27                                       Queen's Park           0.000000   
28                                           Rosedale           0.000000   
29                                           Roselawn           0.000000   
30                                  Runnymede,Swansea           0.000000   
31                            Ryerson,Garden District           0.000000   
32                                     St. James Town           0.000000   
33                    Stn A PO Boxes 25 The Esplanade           0.000000   
34                                    Studio District           0.000000   
35                  The Annex,North Midtown,Yorkville           0.000000   
36                                        The Beaches           0.000000   
37                      The Beaches West,India Bazaar           0.000000   
38                        The Danforth West,Riverdale           0.000000   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0    0.0000              0.0000        0.0000           0.000   
1    0.0000              0.0000        0.0000           0.000   
2    0.0000              0.0000        0.0000           0.000   
3    0.0000              0.0000        0.0000           0.000   
4    0.0625              0.0625        0.0625           0.125   
5    0.0000              0.0000        0.0000           0.000   
6    0.0000              0.0000        0.0000           0.000   
7    0.0000              0.0000        0.0000           0.000   
8    0.0000              0.0000        0.0000           0.000   
9    0.0000              0.0000        0.0000           0.000   
10   0.0000              0.0000        0.0000           0.000   
11   0.0000              0.0000        0.0000           0.000   
12   0.0000              0.0000        0.0000           0.000   
13   0.0000              0.0000        0.0000           0.000   
14   0.0000             

# Let's get the top 5 common venues for each neighbourhood. 

In [60]:
topcan_venues = 5

for hood in torontocan_grouping['Neighbourhood']:
    print("----"+hood+"----")
    temp = torontocan_grouping[torontocan_grouping['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(topcan_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.04
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.03


----Berczy Park----
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.06
2      Beer Bar  0.04
3          Café  0.04
4   Cheese Shop  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0  Breakfast Spot  0.09
1            Café  0.09
2       Nightclub  0.09
3     Coffee Shop  0.09
4   Grocery Store  0.04


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3          Smoke Shop  0.06
4       Burrito Place  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
                venue  freq
0      Airport Lounge  0.12
1     Airport Service  0.12
2    Airport Terminal  0.12
3         Coffee 

# Let's Put all this information in the form of a dataframe.

In [64]:
def common_venues(row, top_venues): # Sorting the venues in Descending Order 
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:top_venues]
                                            
top_venues = 10

indicators = ['st', 'nd', 'rd']
columns = ['Neighbourhood']
                                              
for ind in np.arange(top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))
neighborhoodscan_venues_sorted = pd.DataFrame(columns=columns)
neighborhoodscan_venues_sorted['Neighbourhood'] = torontocan_grouping['Neighbourhood']

for ind in np.arange(torontocan_grouping.shape[0]):
    neighborhoodscan_venues_sorted.iloc[ind, 1:] = common_venues(torontocan_grouping.iloc[ind, :], top_venues)

neighborhoodscan_venues_sorted

Neighbourhood  1st Most Common Venue  \
0                              Adelaide,King,Richmond            Coffee Shop   
1                                         Berczy Park            Coffee Shop   
2          Brockton,Exhibition Place,Parkdale Village            Coffee Shop   
3   Business Reply Mail Processing Centre 969 Eastern     Light Rail Station   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...         Airport Lounge   
5                          Cabbagetown,St. James Town            Coffee Shop   
6                                  Central Bay Street            Coffee Shop   
7             Chinatown,Grange Park,Kensington Market  Vietnamese Restaurant   
8                                            Christie          Grocery Store   
9                                Church and Wellesley            Coffee Shop   
10                      Commerce Court,Victoria Hotel            Coffee Shop   
11                                         Davisville         Sandwich Place   
12                                   Davisville North                   Park   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...                    Pub   
14            Design Exchange,Toronto Dominion Centre            Coffee Shop   
15                        Dovercourt Village,Dufferin                 Bakery   
16              First Canadian Place,Underground city            Coffee Shop   
17                 Forest Hill North,Forest Hill West                   Park   
18                      Harbord,University of Toronto                   Café   
19    Harbourfront East,Toronto Islands,Union Station            Coffee Shop   
20                       High Park,The Junction South        Thai Restaurant   
21  Kingsway Park South West,Mimico NW,The Queensw...            Coffee Shop   
22                                      Lawrence Park                   Park   
23                            Little Portugal,Trinity                    Bar   
24                         Moore Park,Summerhill East             Playground   
25                                 North Toronto West         Clothing Store   
26                              Parkdale,Roncesvalles              Gift Shop   
27                                       Queen's Park            Coffee Shop   
28                                           Rosedale                   Park   
29                                           Roselawn                 Garden   
30                                  Runnymede,Swansea                   Café   
31                            Ryerson,Garden District            Coffee Shop   
32                                     St. James Town            Coffee Shop   
33                    Stn A PO Boxes 25 The Esplanade            Coffee Shop   
34                                    Studio District                   Café   
35                  The Annex,North Midtown,Yorkville         Sandwich Place   
36                                        The Beaches                  Trail   
37                      The Beaches West,India Bazaar                   Park   
38                        The Danforth West,Riverdale       Greek Restaurant   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0             Steakhouse                  Café                          Bar   
1           Cocktail Bar              Beer Bar                       Bakery   
2         Breakfast Spot                  Café                    Nightclub   
3            Yoga Studio         Auto Workshop                         Park   
4        Airport Service      Airport Terminal              Harbor / Marina   
5     Italian Restaurant                  Park                  Pizza Place   
6         Ice Cream Shop    Italian Restaurant                    Juice Bar   
7                    Bar                  Café                  Coffee Shop   
8                   Café                  Park                    Nightclub   
9    Japanese Restaurant      Sushi Restaurant   

# Applying K-Means Clustering to create Clusters of the neighbourhoods.  

In [65]:
kclusters = 5 # Defining the Number of Clusters

torontocan_grouping_kmeansclustering = torontocan_grouping.drop('Neighbourhood', 1)
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(torontocan_grouping_kmeansclustering)
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

# Let's Create a New DataFrame that combines the previous dataframe and the cluster as follows. 

In [70]:
torontocan_merger = toronto_can 
torontocan_merger['Clustering Labels'] = kmeans.labels_ 
torontocan_merger = torontocan_merger.join(neighborhoodscan_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')
torontocan_merger.head()

PostalCode          Borough                  Neighbourhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Clusteriing Labels  Clustering Labels 1st Most Common Venue  \
0 -79.293031                   0                  0                 Trail   
1 -79.352188                   0                  0      Greek Restaurant   
2 -79.315572                   0                  0                  Park   
3 -79.340923                   0                  0                  Café   
4 -79.388790                   0                  0                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Health Food Store           Pizza Place                   Pub   
1           Coffee Shop    Italian Restaurant            Restaurant   
2           Pizza Place                   Pub        Sandwich Place   
3           Coffee Shop             Gastropub                Bakery   
4           Swim School              Bus Line                Lawyer   

  5th Most Common Venue    6th Most Common Venue    7th Most Common Venue  \
0          Dessert Shop       Dim Sum Restaurant                    Diner   
1        Ice Cream Shop   Furniture / Home Store  Fruit & Vegetable Store   
2          Burger Joint       Italian Restaurant            Burrito Place   
3    Italian Restaurant                  Brewery      American Restaurant   
4            Comic Shop  Comfort Food Restaurant     Ethiopian Restaurant   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0        Discount Store                      Dog Run            Yoga Studio  
1                   Pub                  Pizza Place           Liquor Store  
2  Fast Food Restaurant                 Liquor Store         Ice Cream Shop  
3           Yoga Studio           Seafood Restaurant         Sandwich Place  
4     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

# The Columns Clsuteriing Labels and Clustering Labels are the same.     

# Let's Visualize the Clusters.

In [73]:
torontocan_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
markers_colors = []
for lat, lon, poi, cluster in zip(torontocan_merger['Latitude'], torontocan_merger['Longitude'], torontocan_merger['Neighbourhood'], torontocan_merger['Clustering Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(torontocan_clusters)
       
torontocan_clusters

# Examining the Cluster.

In [74]:
torontocan_merger.loc[torontocan_merger['Clustering Labels'] == 0, torontocan_merger.columns[[1] + list(range(5, torontocan_merger.shape[1]))]]

Borough  Clusteriing Labels  Clustering Labels  \
0       East Toronto                   0                  0   
1       East Toronto                   0                  0   
2       East Toronto                   0                  0   
3       East Toronto                   0                  0   
4    Central Toronto                   0                  0   
5    Central Toronto                   0                  0   
6    Central Toronto                   0                  0   
7    Central Toronto                   0                  0   
8    Central Toronto                   0                  0   
9    Central Toronto                   0                  0   
10  Downtown Toronto                   0                  0   
11  Downtown Toronto                   0                  0   
12  Downtown Toronto                   0                  0   
13  Downtown Toronto                   0                  0   
14  Downtown Toronto                   0                  0   
15  Downtown Toronto                   0                  0   
16  Downtown Toronto                   0                  0   
18  Downtown Toronto                   0                  0   
19  Downtown Toronto                   0                  0   
20  Downtown Toronto                   0                  0   
21  Downtown Toronto                   0                  0   
23   Central Toronto                   0                  0   
25  Downtown Toronto                   0                  0   
26  Downtown Toronto                   0                  0   
27  Downtown Toronto                   0                  0   
30  Downtown Toronto                   0                  0   
31      West Toronto                   0                  0   
32      West Toronto                   0                  0   
33      West Toronto                   0                  0   
34      West Toronto                   0                  0   
35      West Toronto                   0                  0   
36      West Toronto                   0                  0   
37  Downtown Toronto                   0                  0   
38      East Toronto                   0                  0   

    1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                   Trail     Health Food Store           Pizza Place   
1        Greek Restaurant           Coffee Shop    Italian Restaurant   
2                    Park           Pizza Place                   Pub   
3                    Café           Coffee Shop             Gastropub   
4                    Park           Swim School              Bus Line   
5                    Park                 Hotel        Breakfast Spot   
6          Clothing Store           Coffee Shop   Sporting Goods Shop   
7          Sandwich Place           Pizza Place          Dessert Shop   
8              Playground           Summer Camp           Yoga Studio   
9                     Pub           Coffee Shop           Pizza Place   
10                   Park            Playground                 Trail   
11            Coffee Shop    Italian Restaurant                  Park   
12            Coffee Shop   Japanese Restaurant      Sushi Restaurant   
13            Coffee Shop                  Park                Bakery   
14            Coffee Shop        Clothing Store        Cosmetics Shop   
15            Coffee Shop                  Café            Restaurant   
16            Coffee Shop          Cocktail Bar              Beer Bar   
18            Coffee Shop            Steakhouse                  Café   
19            Coffee Shop              Aquarium    Italian Restaurant   
20            Coffee Shop                  Café                 Hotel   
21            Coffee Shop                  Café                 Hotel   
23                   Park         Jewelry Store                 Trail   
25                   Café                   Bar             Bookstore   
26  Vietnamese Restaurant                   Bar        

# PART 3  Ends Here.